### ISSUE TITLE
Are the performance metrics reliable?

### ISSUE THREAD

**Issue by hkup859** [sigmaSd/Aoc2024Bench#3]
I haven't dug into the way this project is testing/analyzing performance yet, but when I was playing around with it locally I noticed some odd behavior. I uploaded the same solution on accident under a different name, and I got pretty different results (2.39ms vs 6.15ms). I tested again before creating this issue and it gave me 3.76ms, all for the same code.

Is the scoring/code being run 1 time per solution submission? If so, should the code be run several times (ideally thousands) to get a more reliable average?

I haven't tested this against the live leaderboard at https://aoc2024bench.deno.dev/ because I didn't want to clutter the leaderboard with identical solutions, but testing in a deployed environment may yield different results than local, especially in terms of consistency.

**Comment by sigmaSd** [comment]
The current way it tests it just inserts performance.now before and after your code and run it in the server

This means that if the server is busy doing something else like evaluating concurrently other solutions or maybe just simple fetch requests it might affect the result

I considered moving the evaluation to the client ,but that would make it dependent on the users hardware

i have one feature i thought its good to add, which is only evaluate one code at a time on the server, the other submitters wait on a queue

I think this combined with your idea of running the code multiple time and averiging it, would make this way more reliable

TLDR; missing features
- Run one solution on the server at a time, make other users wait on a queue 
- Run the code multiple time and give average

**Comment by hkup859** [comment]
Sounds good. Another thing to factor in with running multiple times is caching. When testing locally (just some basic code to run the solution 10k times), the time trends downward until it reaches what I assume is the minimum run time. To mitigate this we could throw out the first x number of runs, choose the fastest run only, and or have a very large amount of runs (though that gets pretty slow & intensive).

From my basic testing, we could probably throw out the first 100 runs out of 10k, or implement some logic to remove the best and worst 1% (or 5%) to get a better idea of true average.

Just brainstorming at this point, do you have any preferred solutions? Happy to jump in and help code it.


I also agree that running only 1 solution at a time is necessary to ensure fairness in testing.

**Comment by sigmaSd** [comment]
I think https://github.com/sharkdp/hyperfine have already these statistics tricks, we can look there for more ideas 

I think we can start simple any improvement is welcome, feel free to send a pr, just make sure the whole run don't exceed a certain amount of time 

The queueing system is more involved if you think you can move this forward feel free to also open a pr otherwise I can tackle this when I have more Free time

**Comment by hkup859** [comment]
Is there a certain time limit you have in mind? I know the time limit is currently set to 5 seconds per run. Perhaps something like 5 minutes total? Or is that too ambitious?

I'll take a look at the hyperfine statistics tricks to gather some ideas. With what I've looked at so far I think I'll start on the multiple runs first. The queueing system should be pretty straight forward once a database mechanism is chosen. It could be something simple like using the filesystem or system memory. Alternatively (and a bit more robust) we could deploy a DB alongside the server, but I'm not sure what your deploy setup is like. There might be a free online option to use to avoid getting too complex in this project. Something like DynamoDB would be simple to use and free so long as we don't exceed the storage limit of the free tier.

**Comment by sigmaSd** [comment]
I'm already using Deno.Kv as the database ßince it just works with deno deploy

**Comment by hkup859** [comment]
Good to know! Will look into that.

Another way to improve accurate performance metrics would be to increase the input sizes. How are you currently generating the input text files?

**Comment by sigmaSd** [comment]
I'm just copying the input I get in my own AOC 2024 run

**Comment by hkup859** [comment]
Ah, I never realized the inputs were different per user, lol. That makes sense.

**Comment by hkup859** [comment]
Adding a note:

While I'm still kicking around some ideas for queueing and improving the reliability of the performance metrics, v1.2 is reasonably accurate so far. Hopefully if we add in some queueing or some other method of increasing the total number of runs then we can re-run all current solutions against the latest version to get comparable numbers.

**Comment by sigmaSd** [comment]
I add you as a collaborator for this project, its a project for fun either way, feel free to improve the current state and I'll try as well when I get some free time, thanks

**Comment by hkup859** [comment]
Thanks! I've been having fun testing with it and recently promoted it in some Discord channels, so hopefully we'll get some more submissions.
